File: server/Models/SystemMetrics.cs
using System.Collections.Generic;

namespace EagleEyeMonitor.Server.Models;

public class SystemMetrics
{
    public CpuMetrics Cpu { get; set; } = new();
    public MemoryMetrics Memory { get; set; } = new();
    public DiskMetrics Disk { get; set; } = new();
    public NetworkMetrics Network { get; set; } = new();
    public GpuMetrics Gpu { get; set; } = new();
    public List<ProcessInfo> TopProcesses { get; set; } = new();
    public List<MemoryProcess> TopMemoryProcesses { get; set; } = new();
    public MemoryDistribution MemoryDistribution { get; set; } = new();
    public PerformanceHistory History { get; set; } = new();
    public int Processes { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}

public class CpuMetrics
{
    public double Usage { get; set; }
    public double Temperature { get; set; }
    public double Frequency { get; set; }
    public int Cores { get; set; }
    public int Threads { get; set; }
    public List<CpuCoreMetrics> CoreData { get; set; } = new();
}

public class CpuCoreMetrics
{
    public string Core { get; set; } = string.Empty;
    public double Usage { get; set; }
    public double Temperature { get; set; }
}

public class MemoryMetrics
{
    public double Usage { get; set; }
    public double Total { get; set; }
    public double Available { get; set; }
    public MemoryDetail Physical { get; set; } = new();
    public MemoryDetail Virtual { get; set; } = new();
    public MemoryDetail Cache { get; set; } = new();
    public MemoryDetail Swap { get; set; } = new();
}

public class DiskMetrics
{
    public double Usage { get; set; }
    public double ReadSpeed { get; set; }
    public double WriteSpeed { get; set; }
}

public class NetworkMetrics
{
    public double Usage { get; set; }
    public double UploadSpeed { get; set; }
    public double DownloadSpeed { get; set; }
}

public class GpuMetrics
{
    public double Usage { get; set; }
    public double Temperature { get; set; }
    public double MemoryUsed { get; set; }
    public double MemoryTotal { get; set; }
    public double Power { get; set; }
    public double FanSpeed { get; set; }
}

public class ProcessInfo
{
    public string Name { get; set; } = string.Empty;
    public int Pid { get; set; }
    public double Usage { get; set; }
    public double Memory { get; set; }
}

public class MemoryProcess
{
    public string Name { get; set; } = string.Empty;
    public double Memory { get; set; }
    public string Type { get; set; } = string.Empty;
}

public class MemoryDetail
{
    public double Used { get; set; }
    public double Total { get; set; }
    public double Percentage { get; set; }
}

public class MemoryDistribution
{
    public double Applications { get; set; }
    public double System { get; set; }
    public double Cache { get; set; }
    public double Free { get; set; }
}

public class PerformanceHistory
{
    public List<double> Cpu { get; set; } = new();
    public List<double> CpuTemperature { get; set; } = new();
    public List<double> CpuFrequency { get; set; } = new();
    public List<double> Memory { get; set; } = new();
    public List<double> Disk { get; set; } = new();
    public List<double> Network { get; set; } = new();
    public List<double> Gpu { get; set; } = new();
}

File: src/Types/systemMetrics.ts
export type SystemMetrics = {
  cpu: {
    usage: number;
    temperature: number;
    frequency: number;
    cores: number;
    coreData: { core: string; usage: number; temperature: number }[];
  };
  memory: {
    usage: number;
    total: number;
    available: number;
    physical: { used: number; total: number; percentage: number };
    virtual: { used: number; total: number; percentage: number };
    cache: { used: number; total: number; percentage: number };
    swap: { used: number; total: number; percentage: number };
  };
  disk: {
    usage: number;
    readSpeed: number;
    writeSpeed: number;
  };
  network: {
    usage: number;
    uploadSpeed: number;
    downloadSpeed: number;
    ping: number;
    data: { up: number; down: number };
    interfaces: { name: string; status: string; speed: string; ip: string }[];
    connections: {
      protocol: string;
      remoteAddress: string;
      port: number;
      state: string;
      process: string;
    }[];
  };
  gpu: {
    usage: number;
    temperature: number;
    memoryUsed: number;
    memoryTotal: number;
    power: number;
    fanSpeed: number;
  };
  processes: number;
  topProcesses: {
    name: string;
    pid: number;
    usage: number;
    memory: number;
  }[];
  topMemoryProcesses: { name: string; memory: number; type: string }[];
  memoryDistribution: {
    applications: number;
    system: number;
    cache: number;
    free: number;
  };
  history: {
    cpu: number[];
    cpuTemperature: number[];
    cpuFrequency: number[];
    memory: number[];
    disk: number[];
    network: number[];
    networkDownload: number[];
    networkUpload: number[];
    ping: number[];
    gpu: number[];
  };
  timestamp: string;
};
